grpc-api/src/main/java/io/deephaven/grpc_api/session/SessionState.java

```diff
@@ -289,12 +292,24 @@ public AuthContext getAuthContext() {
     }
 
     /**
-     * Attach a non-export to the liveness scope bound to the session.
+     * Attach an on-close callback bound to the life of the session.
      *
-     * @param referent the object to manage
+     * @param onClose the callback to invoke at end-of-life
      */
-    public void attachLivenessReferent(final LivenessReferent referent) {
-        nonExportsToClose.manage(referent);
+    public synchronized void addOnCloseCallback(final Closeable onClose) {
```

Comment:
This method should be synchronized on `onCloseCallbacks`, not on `this`.
At that point, I would just switch to synchronizing the remove and clear, too, and changing the reference manager to "not concurrent".

Reply:
I would use this pattern for exportListeners, too, and back off to an `ArrayList` for those.